overlord/snapshotstate/snapshotstate.go

```diff
@@ -414,11 +414,20 @@ func Save(st *state.State, instanceNames []string, users []string) (setID uint64
 	for _, name := range instanceNames {
 		desc := fmt.Sprintf("Save data of snap %q in snapshot set #%d", name, setID)
 		task := st.NewTask("save-snapshot", desc)
+
+		opts := options[name]
+		snapshotOptions := &opts
+		if snapshotOptions.IsEmpty() {
+			snapshotOptions = nil // Omit empty options
+		}
```

Comment:
I don't think it makes sense for no paths to be represented as a zero value struct. In particular since we're using a map, if there are no options, we could just not have that mapping at all. Then we wouldn't need this check which seems a bit strange

Reply:
Thanks for this comment. 

If I understand you correctly, I think this ties in with [Michael's related comment](https://github.com/snapcore/snapd/pull/12584/files/ba7695a68afe5ee2591d26cddb51a83ad8ee3c8e#r1114723611).

`snapshotOptions` will be inserted into the field of a structure that is marshalled into snapshot and API response metadata. The only way to insert nothing is with nil pointer to object.

The reality on the unmarshalling side is that there are permutation of valid json that would lead to non-nil pointer to "empty" structure. This leads to `"options":{}` ending up in metadata.

So I opted for an explicit check for "empty" (container of empty containers is empty). This cannot be avoided even with map to pointers to objects as explained in response to [Michael's related comment](https://github.com/snapcore/snapd/pull/12584/files/ba7695a68afe5ee2591d26cddb51a83ad8ee3c8e#r1114723611).

As I said to Michael, its perhaps better general rule and more symmetrical to use pointers to structure for unmarshalling object, I just avoided 1 extra line of code to check for nil :)

Did I understand you correctly? What on your toughs on rather using map to pointers?

Reply:
See new approach with custom unmarshalling of options map (map to pointers to options).

Reply:
I'm not sure that we're talking about the same thing, or maybe I just missed your point. What I'm saying is that it seems strange to me that `options map[string]snap.SnapshotOptions` has a mapping for each snap to an options struct which may even be a zero value. I'd assume that there should only be mappings for snaps for which the API request had options. If the API request doesn't have any options, that would be unmarshaled as a nil map. If it has options for some of the snaps, the map would only have those mappings. So my question is why does this code assume that the map has options for each snap and then check if the fields of the option are unset or not?

Reply:
"If the API request doesn't have any options, that would be unmarshaled as a nil map" _- Agreed_
"If it has options for some of the snaps, the map would only have those mappings." _- Agreed*, but there is important detail about the definition of "has options" from the perspective of the default unmarshaler._

I am trying to prevent incomplete entries from causing options structure to be marshalled into metadata. 
As an example: 
```
{
   "options": {
       "snap1": {}
   }
}
```  
The default unmarshal would result in non-nil map entry (len(map) is 1) with structure `options` containing empty slice `Exclude []`. This would be marshalled to metadata as `"options": {}` which is what I am trying to avoid. There are other permutations.

Reply:
This was discussed. Currently solution seems reasonable and will receive closer look on re-review.